### PR TITLE
CEP 57: Fast Sync

### PR DESCRIPTION
**Pull request by xcthulhu**
[Rendered][1]

[1]: https://github.com/casper-network/ceps/blob/7d5e2a9cf1f9de6007ad912b284ee062e5be4f71/text/0075-fast-sync.md

### CODE REVIEW

File: text/0075-fast-sync.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Fast syncing nodes use a trusted hash to download the database in a
+secure manner. The trusted hash enables secure download of the latest
+block header. That header contains the hash of the database state. Using
+that database hash, nodes download the database state in a
+BitTorent-like manner. After that nodes execute blocks. When they have
```
Comment (afck) [inline text/0075-fast-sync.md:39]:
```suggestion
BitTorrent-like manner. After that nodes execute blocks. When they have
```

File: text/0075-fast-sync.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    // Download the block chain database (aka the trie-store)
+    let mut outstanding_trie_hashes: Vec<Digest> = vec![current_header.state_root];
+    while let Some(trie_hash) = outstanding_trie_hashes.pop() {
+        let trie: Trie<Key, StoredValue> = fetch_trie(trie_hash).await;
+        let new_outstanding_trie_hashes =
+            put_trie_and_find_missing_descendant_trie_hashes(trie).await;
+        outstanding_trie_hashes.extend(new_outstanding_trie_hashes);
+    }
+
+    // Download and execute blocks to current
```
Comment (afck) [inline text/0075-fast-sync.md:98]: Maybe add a comment why we're doing this: Downloading the trie store could have taken a long time, so the current `current_header` might not be so current anymore.

**Review by afck — APPROVED** [review]

File: text/0075-fast-sync.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+Nodes only need recent database states to operate, so it saves disk
+space to fast sync.
+
+Fast syncing avoids the need to be backward compatible with old
+versions. This is because the current executable may not support
+executing old blocks. New versions need to support old executables
+downloading blocks.
+
+After implementation nodes will join the network by fast syncing.
+Operators will only need the current to join the network.
```
Comment (fizyk20) [inline text/0075-fast-sync.md:29]:
```suggestion
Operators will only need the current version to join the network.
```

**Review by fizyk20 — APPROVED** [review]

File: text/0075-fast-sync.md (hunk 3 above)
Comment (afck) [reply text/0075-fast-sync.md]: Done in https://github.com/casper-network/ceps/pull/75/commits/a43ffc357c3c606c0cf7cfe65f5150c1bb408b46.

File: text/0075-fast-sync.md (hunk 2 above)
Comment (afck) [reply text/0075-fast-sync.md:98]: Done in https://github.com/casper-network/ceps/pull/75/commits/a43ffc357c3c606c0cf7cfe65f5150c1bb408b46.

**Comment by cspramit** [comment]
CEP does not talk about archival nodes

**Comment by afck** [comment]
@cspramit: There seems to be a separate CEP, https://github.com/casper-network/ceps/pull/66, for archival sync.

**Comment by xcthulhu** [comment]
I went ahead and merged this without seeing you had tagged other people, @afck. If there are any other review comments I can make a PR with updates.

**Pull request merged**